core/node/core.go

```diff
@@ -152,63 +166,73 @@ func Dag(bs blockservice.BlockService) format.DAGService {
 }
 
 // Files loads persisted MFS root
-func Files(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo, dag format.DAGService, bs blockstore.Blockstore) (*mfs.Root, error) {
-	dsk := datastore.NewKey("/local/filesroot")
-	pf := func(ctx context.Context, c cid.Cid) error {
-		rootDS := repo.Datastore()
-		if err := rootDS.Sync(ctx, blockstore.BlockPrefix); err != nil {
-			return err
-		}
-		if err := rootDS.Sync(ctx, filestore.FilestorePrefix); err != nil {
-			return err
+func Files(reprovidingStrategy string) func(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo, dag format.DAGService, bs blockstore.Blockstore, prov provider.System) (*mfs.Root, error) {
+	return func(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo, dag format.DAGService, bs blockstore.Blockstore, prov provider.System) (*mfs.Root, error) {
+		dsk := datastore.NewKey("/local/filesroot")
+		pf := func(ctx context.Context, c cid.Cid) error {
+			rootDS := repo.Datastore()
+			if err := rootDS.Sync(ctx, blockstore.BlockPrefix); err != nil {
+				return err
+			}
+			if err := rootDS.Sync(ctx, filestore.FilestorePrefix); err != nil {
+				return err
+			}
+
+			if err := rootDS.Put(ctx, dsk, c.Bytes()); err != nil {
+				return err
+			}
+			return rootDS.Sync(ctx, dsk)
 		}
 
-		if err := rootDS.Put(ctx, dsk, c.Bytes()); err != nil {
-			return err
-		}
-		return rootDS.Sync(ctx, dsk)
-	}
-
-	var nd *merkledag.ProtoNode
-	ctx := helpers.LifecycleCtx(mctx, lc)
-	val, err := repo.Datastore().Get(ctx, dsk)
-
-	switch {
-	case err == datastore.ErrNotFound || val == nil:
-		nd = unixfs.EmptyDirNode()
-		err := dag.Add(ctx, nd)
-		if err != nil {
-			return nil, fmt.Errorf("failure writing filesroot to dagstore: %s", err)
-		}
-	case err == nil:
-		c, err := cid.Cast(val)
-		if err != nil {
+		var nd *merkledag.ProtoNode
+		ctx := helpers.LifecycleCtx(mctx, lc)
+		val, err := repo.Datastore().Get(ctx, dsk)
+
+		switch {
+		case err == datastore.ErrNotFound || val == nil:
```

Comment:
Checking `val == nil` seems wrong, because `val` should be `nil` for any non-nil error that is returned. Also, it would be better to use `errors.Is` instead of direct comparison.

```suggestion
		case errors.Is(err, datastore.ErrNotFound):
```

Reply:
This code is just re-indented I think, it's been like that for 6 years https://github.com/ipfs/kubo/commit/d35dac70f0b0b47c7813c6380deb8974179bdb32. 

I'll take your proposal though.